targets/ublk.nfs.cpp

```diff
@@ -0,0 +1,383 @@
+/* -*-  mode:c; tab-width:8; c-basic-offset:8; indent-tabs-mode:nil;  -*- */
+// SPDX-License-Identifier: MIT or GPL-2.0-only
+
+#include <config.h>
+
+#include "ublksrv_tgt.h"
+#include <nfsc/libnfs.h>
+
+
+struct nfs_tgt_data {
+        char url[4096];
+	struct nfs_context *nfs;
+        struct nfsfh *nfsfh;
+        size_t capacity;
+};
+
+typedef struct nfs_cb_data {
+        struct nfs_cb_data *next;
+        const struct ublksrv_queue *q;
+        int tag;
+        ssize_t count;
+} nfs_cb_data_t;
+
+nfs_cb_data_t *io_list;
+
+pthread_spinlock_t io_list_lock;
+
+static void nfs_handle_event(const struct ublksrv_queue *q)
+{
+        nfs_cb_data_t *cb_data, *tmp;
+        
+	pthread_spin_lock(&io_list_lock);
+        cb_data = io_list;
+        io_list = NULL;
+	pthread_spin_unlock(&io_list_lock);
+
+        while (cb_data) {
+                tmp = cb_data->next;
+
+                ublksrv_complete_io(cb_data->q, cb_data->tag, cb_data->count);
+                free(cb_data);
+                cb_data = tmp;
+        }
+        ublksrv_queue_handled_event(q);
+}
+
+void rw_async_cb(int status, struct nfs_context *nfs,
+                 void *data, void *private_data)
+{
+        nfs_cb_data_t *cb_data = (nfs_cb_data_t *)private_data;
+
+        if (status < 0) {
+                fprintf(stderr, "pread/pwrite failed with \"%s\"\n", (char *)data);
+                status = -EIO;
+        }
+        cb_data->count = status;
+        
+	pthread_spin_lock(&io_list_lock);
+        cb_data->next = io_list;
+        io_list = cb_data;
+	pthread_spin_unlock(&io_list_lock);
+        
+        ublksrv_queue_send_event(cb_data->q);
+}
```

Comment:
with per-queue nfs data,  sending event can be optimized by just sending it for 1st completed IO before it is completed. IOs are often completed in batch, so no need to send event for other completed IOs
if they are not handled/completed yet.

Reply:
I don't follow. How would I do this from the callback?

Reply:
Done.   I assume calling ublksrv_queue_send_event() from under a spinlock is safe
as we should never get to the point that the write here will block.